server/src/main/java/org/opensearch/telemetry/tracing/listener/TraceableSearchRequestOperationsListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.listener;
+
+import org.opensearch.action.search.SearchPhaseContext;
+import org.opensearch.action.search.SearchRequestContext;
+import org.opensearch.action.search.SearchRequestOperationsListener;
+import org.opensearch.telemetry.tracing.AttributeNames;
+import org.opensearch.telemetry.tracing.Span;
+import org.opensearch.telemetry.tracing.SpanBuilder;
+import org.opensearch.telemetry.tracing.SpanContext;
+import org.opensearch.telemetry.tracing.SpanScope;
+import org.opensearch.telemetry.tracing.Tracer;
+import org.opensearch.telemetry.tracing.noop.NoopSpan;
+
+import static org.opensearch.core.common.Strings.capitalize;
+
+/**
+ * SearchRequestOperationsListener subscriber for search request tracing
+ *
+ * @opensearch.internal
+ */
+public final class TraceableSearchRequestOperationsListener extends SearchRequestOperationsListener {
+    private final Tracer tracer;
+    private Span requestSpan;
```

Comment:
I believe this is not going to work, you have many phases and the phaseXxx events will be called per all or some of them, there would be many spans, not one. I believe you need to keep a map of phase <-> span to track that

Reply:
We are creating a new `TraceableSearchRequestOperationsListener` per request. Each request will only have **one** requestSpan and **one** phaseSpan **at a time**.

Reply:
I am pretty sure this is not right assumption (for phase), check `SearchDfsQueryThenFetchAsyncAction` fe please

Reply:
At the coordinator level there is no overlap. Phases ex. `SearchDfsQueryThenFetchAsyncAction` have a `getNextPhase()` method. Whenever a phase is about to end and `getNextPhase()` is called, we end the current phase with `onPhaseEnd()` then start the next phase with `onPhaseStart()`

Reply:
Oh ... sorry, got it now: you mean there is only one phase active (missed that  **at a time**). Could you please add asserts to that in the listener?

Reply:
The SearchRequestOperationsListener listener framework itself does not maintain any state so we'd have to rely on the subscriber TraceableSearchRequestOperationsListener. But IIRC there is no way to check if a Span is closed. We can create a new flag or set `phaseSpan` to `null`/`NoopSpan` after closing. Any other ideas?

Based on the code, we should be confident that only one search phase occurs at a time and all phases are "ended" by either `onPhaseEnd()` or `onPhaseFailure()`. Is this assertion really necessary?
```
protected void onPhaseEnd(SearchPhaseContext context, SearchRequestContext searchRequestContext) {
    phaseSpan.endSpan();
    phaseSpan = null; // <-- add this
    phaseSpanScope.close();
}
```

Reply:
I was thinking about something like that (so everyone would understand why we have only one span even if there could be many phases):

```
   @Override
    protected void onPhaseStart(SearchPhaseContext context) {
       assert phaseSpan == null ? "There should be only one search phase active at a time"
        phaseSpan = tracer.startSpan( ... );
```

Reply:
Added
```
assert phaseScopedSpan == null : "There should be only one search phase active at a time";
```